src/cuttlefish_generator.erl

```diff
@@ -45,11 +45,75 @@ map(Schema, Config) ->
 
 map(Schema, Config, ConfFile) ->
     IncludeFile = lists:filter(fun({K, _}) -> K =:= include_file end, Config),
-    AllConfig = merge_include_conf(Config -- IncludeFile, IncludeFile, ConfFile),
-    map_add_defaults(Schema, AllConfig).
+    IncludedConfig = merge_include_conf(Config -- IncludeFile, IncludeFile, ConfFile),
+    try
+        AllConfig = merge_env_conf(IncludedConfig, Schema),
+        map_add_defaults(Schema, AllConfig)
+    catch
+        error:Reason -> {error, map, [{errorlist, [{error, Reason}]}]}
+    end.
+
+-spec (merge_env_conf(cuttlefish_conf:conf(), cuttlefish_schema:schema()) -> cuttlefish_conf:conf()).
+merge_env_conf(Config, Schema) ->
+    Envs = match_env(getenv(), Schema),
+    lists:ukeymerge(1, Envs, Config).
+
+-spec (getenv() -> cuttlefish_conf:conf()).
+getenv() ->
+    lists:map(fun (E) -> [Key, Value] = string:split(E, "=", leading), {Key, Value} end, os:getenv()).
+
+match_env(AllEnvs, {_, Mappings, _}) ->
+    Prefix = os:getenv("CUTTLEFISH_ENV_OVERRIDE_PREFIX", undefined),
+    Envs = lists:filtermap(fun (KV) -> check_prefix(KV, Prefix) end, AllEnvs),
+    Default = lists:ukeysort(1, match_default_env(Envs, Mappings)),
+    Override = lists:ukeysort(1, match_override_env(Envs, Mappings)),
+    lists:ukeymerge(1, Override, Default).
+
+match_default_env(Envs, Mappings) ->
+    lists:filtermap(fun (KV) -> do_match_default_env(KV, Mappings) end, Envs).
+
+do_match_default_env({Key, Value}, Mappings) ->
+    FormattedKey = string:lowercase(string:replace(Key, "__", ".", all)),
+    T = cuttlefish_variable:tokenize(FormattedKey),
+    case lists:any(fun (M) -> cuttlefish_variable:is_fuzzy_match(T, cuttlefish_mapping:variable(M)) end, Mappings) of
+        true ->
+            {true, {T, Value}};
+        false ->
+            false
+    end.
+
+check_prefix(_, undefined) ->
+    false;
+check_prefix({Key, Value}, Prefix) ->
+    case string:prefix(Key, Prefix) of
+        nomatch ->
+            false;
+        Remainder ->
+            {true, {Remainder, Value}}
+    end.
+
+match_override_env(Envs, Mappings) ->
+    lists:foldl(fun (KV, A) -> lists:append(do_match_override_env(KV, Mappings), A) end, [], Envs).
+
+do_match_override_env({Key, Value}, Mappings) ->
+    do_match_override_env({Key, Value}, Mappings, []).
+do_match_override_env(_, [], Acc) ->
+    Acc;
+do_match_override_env({Key, Value}=KV, [M|More], Acc) ->
+    case {Key =:= cuttlefish_mapping:override_env(M), is_fuzzy_variable(cuttlefish_mapping:variable(M))} of
+        {true, true} ->
+            error({no_env_override_support_for_fuzzy_mappings, M});
+        {true, false} ->
+            do_match_override_env(KV, More, [{cuttlefish_mapping:variable(M), Value}|Acc]);
+        {false, _} ->
+            do_match_override_env(KV, More, Acc)
+    end.
+
+is_fuzzy_variable(Variable) ->
+    lists:any(fun ([$$|_]) -> true; (_) -> false end, Variable).
 
 merge_include_conf(Config, [], _ConfFile) ->
-    Config;
+    lists:sort(Config);
```

Comment:
is this only to serve tests ?

Reply:
or it's leftover from before when the code relies on sorted lists to match variable names ?

Reply:
No. When there is no include, merge_include_conf returns unsorted config. This was not a problem when this function is called last but now it is.